DPTP-4046: Allow more registry domains for OSD clusters

deepsm007 opened the pull request:
/cc @openshift/test-platform

deepsm007 commented on cmd/ci-secret-bootstrap/main.go:
```diff
@@ -636,6 +636,55 @@ func mutateGlobalPullSecret(original, secret *coreapi.Secret) (bool, error) {
+		Auth: qciToken,
+	}
+	// token is same as QCI
+	quayRepoDomain := api.DomainForService(api.QuayRepo)
```
This might conflict with other secret need to confirm

smg247 commented on cmd/ci-secret-bootstrap/main.go:
```diff
@@ -641,6 +641,29 @@ func mutateGlobalPullSecret(original, secret *coreapi.Secret) (bool, error) {
 		return false, fmt.Errorf("failed to marshal the docker config: %w", err)
 	}
 	original.Data[coreapi.DockerConfigJsonKey] = data
```
This ends up getting overwritten below, I assume the goal is to append to it instead?

deepsm007 replied:
yeah, that is the ideal scenario to append those secrets.

deepsm007 commented on cmd/ci-secret-bootstrap/main.go:
```diff
@@ -541,18 +541,31 @@ func updateSecrets(getters map[string]Getter, secretsMap map[string][]*coreapi.S
+				if err != nil {
+					errs = append(errs, fmt.Errorf("failed to mutate secret %s:%s/%s: %w", cluster, secret.Namespace, secret.Name, err))
+				}
+				if mutated || mutateQCIAPPCIDomain || mutateQuayOpenshiftCI || mutateQCICache {
```
@smg247 I am skeptical about this statement, other than that everything else should be good

deepsm007 commented:
/test e2e

deepsm007 commented:
/test e2e

deepsm007 commented:
/test integration

hongkailiu commented on cmd/ci-secret-bootstrap/main.go:
```diff
@@ -621,27 +621,35 @@ func mutateGlobalPullSecret(original, secret *coreapi.Secret) (bool, error) {
-	registryDomain := api.DomainForService(api.ServiceRegistry)
-	if dockerConfig.Auths == nil || dockerConfig.Auths[api.DomainForService(api.ServiceRegistry)].Auth == "" {
-		return false, fmt.Errorf("failed to get token for %s", registryDomain)
+	if len(dockerConfig.Auths) == 0 {
```
NIT:

```suggestion
	if dockerConfig.Auths == nil {
```

hongkailiu commented on cmd/ci-secret-bootstrap/main.go:
```diff
@@ -621,27 +621,35 @@ func mutateGlobalPullSecret(original, secret *coreapi.Secret) (bool, error) {
+		originalDockerConfig.Auths[domain] = secretbootstrap.DockerAuth{
+			Auth: dockerConfig.Auths[domain].Auth,
+		}
+		if originalToken != originalDockerConfig.Auths[domain].Auth {
```
NIT:

```suggestion
		if !mutatedSecret && originalToken != originalDockerConfig.Auths[domain].Auth {
```

hongkailiu commented on cmd/ci-secret-bootstrap/main.go:
```diff
@@ -621,27 +621,35 @@ func mutateGlobalPullSecret(original, secret *coreapi.Secret) (bool, error) {
-	if err != nil {
-		return false, fmt.Errorf("failed to marshal the docker config: %w", err)
+	if mutatedSecret {
+		data, err := json.Marshal(dockerConfig)
```
```suggestion
		data, err := json.Marshal(originalDockerConfig)
```

This is an important one i guess.
The key here is that we would like to keep the change minimal, i.e., we modify the secret ONLY when it is necessary and we modify only the parts that belong to us.

hongkailiu commented on pkg/api/domain.go:
```diff
@@ -26,6 +26,8 @@ const (
 
 	QCIAPPCIDomain = "quay-proxy.ci.openshift.org"
+
+	QCICache = "qci-pull-through-cache-us-east-1-ci.apps.ci.l2s4.p1.openshiftapps.com/openshift/ci"
```
```suggestion
	QCICacheDomain = "qci-pull-through-cache-us-east-1-ci.apps.ci.l2s4.p1.openshiftapps.com"
```

We will modify https://github.com/openshift/release/blob/master/core-services/ci-secret-bootstrap/_config.yaml#L692 accordingly.

I missed it in https://github.com/openshift/release/pull/52732/files (or I cannot remember the reason now)

hongkailiu reviewed (commented):
Besides the comments from the PR,
I would like to keep the original unit tests unless you can justify the removal.
If you want to add more, it is totally fine with me.

deepsm007 commented:
> Besides the comments from the PR, I would like to keep the original unit tests unless you can justify the removal. If you want to add more, it is totally fine with me.

I have added few more tests and renamed 1 test

deepsm007 commented:
/test e2e

hongkailiu commented on cmd/ci-secret-bootstrap/main_test.go:
```diff
@@ -3101,168 +3103,509 @@ func TestMutateGlobalPullSecret(t *testing.T) {
 			secret: &coreapi.Secret{
 				Data: map[string][]byte{
 					".dockerconfigjson": []byte(`{
-	"auths": {
```
I do not see the change in this section, except the reformatting?
Could we revert them?

hongkailiu commented on cmd/ci-secret-bootstrap/main_test.go:
```diff
@@ -3101,168 +3103,509 @@ func TestMutateGlobalPullSecret(t *testing.T) {
 			expectedErr: fmt.Errorf("failed to parse the original secret: failed to get content from nil secret"),
 		},
 		{
-			name: "basic case: expired auth is replaced",
```
Could we keep this case and rename it to "basic case: auth for app.ci is expired auth is replaced"

hongkailiu replied:
I do not see this one addresse.

hongkailiu commented on cmd/ci-secret-bootstrap/main_test.go:
```diff
@@ -3101,168 +3103,509 @@ func TestMutateGlobalPullSecret(t *testing.T) {
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
 			actual, actualErr := mutateGlobalPullSecret(tc.original, tc.secret)
-			if diff := cmp.Diff(actualErr, tc.expectedErr, testhelper.EquateErrorMessage); diff != "" {
```
Why do we change this section?

deepsm007 replied:
when the slices of registries are not sorted the diff provided is different. Using `cmpopts.SortedSlices()` was also an option but assert seemed more convenient

hongkailiu replied:
Do you mind doing it with another PR?
I would like to keep the change minimal so that the review is easier and the chance of screwing thing up is smaller, in my opinion.

deepsm007 commented:
/test e2e

hongkailiu commented on cmd/ci-secret-bootstrap/main_test.go:
```diff
@@ -3127,142 +3129,457 @@ func TestMutateGlobalPullSecret(t *testing.T) {
 		{
-			name: "basic case: expired auth is replaced",
+			name: "not mutated if the auth is still valid",
+			original: &coreapi.Secret{
```
Could we switch the order back?
If you want you may change it with another PR.

deepsm007 replied:
I ordered them as they were before, new tests are at the end

hongkailiu replied:
The previous order was secret and then original and now it is the other way around.

deepsm007 commented:
/test e2e

deepsm007 commented:
/retest-required

hongkailiu commented:
/lgtm

/hold

Whenever you are ready to watch the rolling AFTER the backup of the secrets is in place.

deepsm007 commented:
/override ci/prow/e2e

deepsm007 commented:
/unhold 

Will watching the rolling

deepsm007 commented:
/override ci/prow/e2e
